README.md
# rpcx

[**中文介绍**](README-CN.md) [**中文电子书**](https://www.gitbook.com/book/smallnest/go-rpc/details)

[![License](https://img.shields.io/:license-apache-blue.svg)](https://opensource.org/licenses/Apache-2.0) [![GoDoc](https://godoc.org/github.com/smallnest/rpcx?status.png)](http://godoc.org/github.com/smallnest/rpcx)  [![travis](https://travis-ci.org/smallnest/rpcx.svg?branch=master)](https://travis-ci.org/smallnest/rpcx) [![Go Report Card](https://goreportcard.com/badge/github.com/smallnest/rpcx)](https://goreportcard.com/report/github.com/smallnest/rpcx) [![QQ群](https://img.shields.io/:QQ群-398044387-blue.svg)](_documents/images/rpcx_qq.png)


rpcx is a distributed RPC framework like [Alibaba Dubbo](http://dubbo.io/) and [Weibo Motan](https://github.com/weibocom/motan).
It is based on Go net/rpc and provides extra governance features.

**Notice**

Since **rpcx 2.0**, we replace underlying Go official rpc transport with customized rpc transport. So rpcx can support **context** in client side and server side. 

If you want to use underlying Go official rpc transport, please use **rpcx 1.3**.


![Throughput](_documents/throughput.png)

very very good performance. Much better than gRPC, Dubbo and weibo Motan.




When we talk about RPC frameworks, Dubbo is the first framework we should mention, and also Dubbox mantained by dangdang.
Dubbo has been widely used in e-commerce companies in China, for example, Alibaba, Jingdong and Dangdang.

Though Dubbo still uses Spring 2.5.6.SEC03, and is no longer supported by Alibaba, 
some other companies still usea it and maintain their own branches.

> DUBBO is a distributed service framework , provides high performance and transparent RPC remote service call. 
> It is the core framework of Alibaba SOA service governance programs. 
> There are 3,000,000,000 calls for 2,000+ services per day, 
> and it has been widely used in various member sites of Alibaba Group.

Motan is open source now by Weibo. As Zhang Lei said, he is the current main developer of Motan:

> Motan started in 2013. There are 100 billion calls for hundreds of service calls every day.

Those two RPC frameworks are developed in Java. 
There are other famous RPC frameworks such as [thrift](https://thrift.apache.org/)、and [finagle](https://github.com/twitter/finagle)。

The goal of [rpcx](https://github.com/smallnest/rpcx) is to implement an RPC framework like Dubbo in Go.
It is developed in Go, and to be used in Go. 

It is a distributed、pluggable RPC framework with governance (service discovery、load balancer、fault tolerance、monitoring, etc.).

As you know, RPC frameworks already exists, for example, [net/rpc](https://golang.org/pkg/net/rpc/)、[grpc-go](https://github.com/grpc/grpc-go)、[gorilla-rpc](http://www.gorillatoolkit.org/pkg/rpc),
So why re-invent the wheel?

Although those Go RPC frameworks work well, their function is relatively simple and they only implement end-to-end communication.
They lack some product features of service management functions like service discovery,
load balancing, and fault tolerance. 

So I created rpcx and expect it to become an RPC framework like Dubbo.


A similar project in go is called [go-micro](https://github.com/micro/go-micro). 


## What's RPC
From wikiPedia:

> In distributed computing, a remote procedure call (RPC) is when a computer program causes a procedure (subroutine) to execute in another address space (commonly on another computer on a shared network), which is coded as if it were a normal (local) procedure call, without the programmer explicitly coding the details for the remote interaction. That is, the programmer writes essentially the same code whether the subroutine is local to the executing program, or remote.[1] This is a form of client–server interaction (caller is client, executer is server), typically implemented via a request–response message-passing system. The object-oriented programming analog is remote method invocation (RMI). The RPC model implies a level of location transparency, namely that calling procedures is largely the same whether it is local or remote, but usually they are not identical, so local calls can be distinguished from remote calls. Remote calls are usually orders of magnitude slower and less reliable than local calls, so distinguishing them is useful.
>
>RPCs are a form of inter-process communication (IPC), in that different processes have different address spaces: if on the same host machine, they have distinct virtual address spaces, even though the physical address space is the same; while if they are on different hosts, the physical address space is different. Many different (often incompatible) technologies have been used to implement the concept.


Sequence of events during an RPC
1. The client calls the client stub. The call is a local procedure call, with parameters pushed on to the stack in the normal way.
2. The client stub packs the parameters into a message and makes a system call to send the message. Packing the parameters is called marshalling.
3. The client's local operating system sends the message from the client machine to the server machine.
4. The local operating system on the server machine passes the incoming packets to the server stub.
5. The server stub unpacks the parameters from the message. Unpacking the parameters is called unmarshalling.
6. Finally, the server stub calls the server procedure. The reply traces the same steps in the reverse direction.


There are two ways to implement RPC frameworks. 
One focusses on cross-language calls and the other focusses on service governance.

Dubbo、DubboX、and Motan are RPC frameworks focusing on service governance.
Thrift、gRPC、Hessian、and Hprose are RPC frameworks focusing on cross-language calls.

rpcx focus in on service governance.

## Features

* Based on net/rpc. a Go net/prc project can be converted to use rpcx with a few changes.
* Pluggable. Features are implemented by Plugins, such as service discovery.
* Communicate with TCP long connections.
* Support GeoLocation
* Support many codec. for example, Gob、Json、MessagePack、gencode、ProtoBuf.
* Service dicovery. support ZooKeeper、Etcd.
* Fault tolerance：Failover、Failfast、Failtry.
* Load banlancer：support randomSelecter, RoundRobin, consistent hashing, etc.
* Scalable.
* Other: metrics、log、timeout.
* Authorization.
* Compression:inflate and snappy.
* Alias name for services.
* [kcp](https://github.com/xtaci/kcp-go) support: a full-featured reliable-UDP library for golang
* support **OpenTracing** and **Prometheus**
* support Header and Context

rpcx has fixed/implemented the below issues in [golang/go rpc](https://github.com/golang/go/issues?utf8=✓&q=is%3Aissue%20rpc) and I believe those issues won't be fixed in the official library because the official rpc library will be frozen.

* [#16449](https://github.com/golang/go/issues/16449): proposal: support custom method names in net/rpc
* [#15236](https://github.com/golang/go/issues/15236): net/rpc: expected Timeout based alternatives to functions for rpc.Dial, rpc.DialHTTP, rpc.DialHTTPPath [proposal].
* [#13395](https://github.com/golang/go/issues/13395): net/rpc: Server.ServeHTTP assumes default http mux
* [#10929](https://github.com/golang/go/issues/10929): net/rpc/jsonrpc: Missing support for JSON-RPC 2.0
* [#7946](https://github.com/golang/go/issues/7946): net/rpc: add client support for RPC over https
* [#4591](https://github.com/golang/go/issues/4591): Authentication for RPC and HTTP
* [#4584](https://github.com/golang/go/issues/4584): net/rpc: access to client's IP address
* [#3615](https://github.com/golang/go/issues/3615): rpc: allow aliasing methods


[rpcx-ui](https://github.com/smallnest/rpcx-ui) provides a web ui for services management.
![rpcx-ui](https://github.com/smallnest/rpcx-ui/blob/master/services.png)

## Architecture
rpcx contains three roles : RPC Server，RPC Client and Registry.
* Server registers services on the Registry
* Client queries the service list and select a server from the server list returned from the Registry.
* When a Server is down, the Registry can remove the server from the list, and subsequently the client can remove it too.

![](https://raw.githubusercontent.com/smallnest/rpcx/master/_documents/images/component.png)

So far rpcx supports ZooKeeper and Etcd as Registry，Consul support is in developement.

## Examples

There is a client calling the Posts service at tr.colobu.com. This service returns a hot tech post list of the day:
```go 
package main

import (
	"encoding/json"
	"fmt"
	"time"

	"gopkg.in/mgo.v2/bson"

	"github.com/smallnest/rpcx"
  "github.com/smallnest/rpcx/log"
)

type Args struct {
	PostType string `msg:"posttype"`
}

type Reply struct {
	Posts []Post `msg:"posts"`
}

type Post struct {
	PostID      bson.ObjectId `json:"id" xml:"id" bson:"_id,omitempty"`
	PostType    string        `json:"ptype" xml:"ptype" bson:"ptype,omitempty"`
	Title       string        `json:"title" xml:"title" bson:"title"`
	URL         string        `json:"url" xml:"url" bson:"url"`
	Domain      string        `json:"domain" xml:"domain" bson:"domain"`
	ShortURL    string        `json:"surl" xml:"surl" bson:"surl"`
	Description string        `json:"desc" xml:"desc" bson:"desc"`
	LikeCount   int           `json:"like" xml:"like" bson:"like"`
	ImageURL    string        `json:"imgurl" xml:"imgurl" bson:"imgurl"`
	RecommendBy string        `json:"-" xml:"-" bson:"-"`
	Tags        string        `json:"tags" xml:"tags" bson:"tags"`
	State       int           `json:"-" xml:"-" bson:"-"`
	Timestamp   time.Time     `json:"ts" xml:"timestamp" bson:"ts"`
}

func main() {
	s := &rpcx.DirectClientSelector{Network: "tcp", Address: "tr.colobu.com:8972", DialTimeout: 10 * time.Second}
	client := rpcx.NewClient(s)
	defer client.Close()

	args := &Args{"golang"}
	var reply Reply
	err := client.Call(context.Background(),"Posts.Query", args, &reply)
	if err != nil {
		log.Infof("error for Posts: %s, %v \n", args.PostType, err)
		return
	}

	posts := reply.Posts
	data, _ := json.MarshalIndent(&posts, "", "\t")

	log.Infof("Posts: %s \n", string(data))
}
```

you can found more examples at [_examples](_examples)



## Benchmark

**Test Environment**
* CPU:    Intel(R) Xeon(R) CPU E5-2620 v2 @ 2.10GHz, 24 cores
* Memory: 16G
* OS:     Linux Server-3 2.6.32-358.el6.x86_64, CentOS 6.4
* Go:     1.7

Test request is copied from protobuf project and encoded to a proto message. Its size is 581 bytes.
The response update two fields of decoded requests so the server goes through decoding and encoding.

The test proto file is:

```proto
syntax = "proto2";

package main;

option optimize_for = SPEED;


message BenchmarkMessage {
  required string field1 = 1;
  optional string field9 = 9;
  optional string field18 = 18;
  optional bool field80 = 80 [default=false];
  optional bool field81 = 81 [default=true];
  required int32 field2 = 2;
  required int32 field3 = 3;
  optional int32 field280 = 280;
  optional int32 field6 = 6 [default=0];
  optional int64 field22 = 22;
  optional string field4 = 4;
  repeated fixed64 field5 = 5;
  optional bool field59 = 59 [default=false];
  optional string field7 = 7;
  optional int32 field16 = 16;
  optional int32 field130 = 130 [default=0];
  optional bool field12 = 12 [default=true];
  optional bool field17 = 17 [default=true];
  optional bool field13 = 13 [default=true];
  optional bool field14 = 14 [default=true];
  optional int32 field104 = 104 [default=0];
  optional int32 field100 = 100 [default=0];
  optional int32 field101 = 101 [default=0];
  optional string field102 = 102;
  optional string field103 = 103;
  optional int32 field29 = 29 [default=0];
  optional bool field30 = 30 [default=false];
  optional int32 field60 = 60 [default=-1];
  optional int32 field271 = 271 [default=-1];
  optional int32 field272 = 272 [default=-1];
  optional int32 field150 = 150;
  optional int32 field23 = 23 [default=0];
  optional bool field24 = 24 [default=false];
  optional int32 field25 = 25 [default=0];
  optional bool field78 = 78;
  optional int32 field67 = 67 [default=0];
  optional int32 field68 = 68;
  optional int32 field128 = 128 [default=0];
  optional string field129 = 129 [default="xxxxxxxxxxxxxxxxxxxxx"];
  optional int32 field131 = 131 [default=0];
}
```


The test is run with 100, 1000, 2000 and 5000 concurrent clients. The total number of requests per clients is 1,000,000.

**Test Result**

### rpcx: one client and one server on the same machine
concurrent clients|mean(ms)|median(ms)|max(ms)|min(ms)|throughput(TPS)
-------------|-------------|-------------|-------------|-------------|-------------
100|0|0|17|0|164338
500|2|1|40|0|181126
1000|4|3|56|0|186219
2000|9|7|105|0|182815
5000|25|22|200|0|178858


you can use test code in `_benchmark` to test.
`server` is used to start a server and `client` is used as clients via protobuf.

The above test is with a client and a server running on the same mechine.

### rpcx: one client and one server running on different machines

If I run them on separate servers, the test results are:

concurrent clients|mean(ms)|median(ms)|max(ms)|min(ms)|throughput(TPS)
-------------|-------------|-------------|-------------|-------------|-------------
100|1|1|20|0|127975
500|5|1|4350|0|136407
1000|10|2|3233|0|155255
2000|17|2|9735|0|159438
5000|44|2|12788|0|161917 

### rpcx: one client on one machine and two servers on two machines

When running in cluster mode, with one machine with one client and another two running as servers, the test results are:

concurrent clients|mean(ms)|median(ms)|max(ms)|min(ms)|throughput(TPS)
-------------|-------------|-------------|-------------|-------------|-------------
100|0|0|41|0|128932
500|3|2|273|0|150285
1000|5|5|621|0|150152
2000|10|7|288|0|159974
5000|23|12|629|0|155279



### benchmarks of serialization libraries:

```
[root@localhost rpcx]# go test -bench . -test.benchmem
PASS
BenchmarkNetRPC_gob-16            100000             18742 ns/op             321 B/op          9 allocs/op
BenchmarkNetRPC_jsonrpc-16        100000             21360 ns/op            1170 B/op         31 allocs/op
BenchmarkNetRPC_msgp-16           100000             18617 ns/op             776 B/op         35 allocs/op
BenchmarkRPCX_gob-16              100000             18718 ns/op             320 B/op          9 allocs/op
BenchmarkRPCX_json-16             100000             21238 ns/op            1170 B/op         31 allocs/op
BenchmarkRPCX_msgp-16             100000             18635 ns/op             776 B/op         35 allocs/op
BenchmarkRPCX_gencodec-16         100000             18454 ns/op            4485 B/op         17 allocs/op
BenchmarkRPCX_protobuf-16         100000             17234 ns/op             733 B/op         13 allocs/op
```

## Comparision with gRPC
[gRPC](https://github.com/grpc/grpc-go) is the RPC framework by Google. It supports multiple programming lanaguage.
I have compared three cases for prcx and gRPC. It shows rpcx is much better than gRPC.

current gRPC version is 1.0.

The test results for rpcx can be found above. Below is the test results for gRPC.

### gRPC: one client and one server on the same machine
concurrent clients|mean(ms)|median(ms)|max(ms)|min(ms)|throughput(TPS)
-------------|-------------|-------------|-------------|-------------|-------------
100|1|0|21|0|68250
500|5|1|3059|0|78486
1000|10|1|6274|0|79980
2000|19|1|9736|0|58129
5000|43|2|14224|0|44724


![](_documents/images/rpcx-grpc-1.png)

### gRPC: one client and one server running on different machines
concurrent clients|mean(ms)|median(ms)|max(ms)|min(ms)|throughput(TPS)
-------------|-------------|-------------|-------------|-------------|-------------
100|1|0|21|0|68250
500|5|1|3059|0|78486
1000|10|1|6274|0|79980
2000|19|1|9736|0|58129
5000|43|2|14224|0|44724 


![](_documents/images/rpcx-grpc-2.png)

### gRPC: one client on one machine and two servers on two machines
concurrent clients|mean(ms)|median(ms)|max(ms)|min(ms)|throughput(TPS)
-------------|-------------|-------------|-------------|-------------|-------------
100|1|0|19|0|88082
500|4|1|1461|0|90334
1000|9|1|6315|0|62305
2000|17|1|9736|0|44487
5000|38|1|25087|0|33198


![](_documents/images/rpcx-grpc-3.png)


CHANGELOG

## v2.0
- replace underlying rpc transport. 替换底层的rpc库，原先使用官方的rpc库，现在在官方的库上做了修改，以便支持context和header
- support opentracing. 支持opentracing协议的tracing系统，比如ZipKin等
- support Prometheus. 支持 Prometheus, 当前比较火的监控系统和时间序列库
- support Header. 客户端可以设置Header,服务器端可以获取这个header
- support context. 客户端可以设置context，通过context可以取消调用。服务器端方法定义的第一个参数可以是context,通过它能获取客户端传递的header信息和客户端的net.Conn

server.go
package rpcx

import (
	"context"
	"crypto/tls"
	"io"
	"net"
	"net/http"
	"regexp"
	"strings"
	"time"

	msgpackrpc2 "github.com/rpcx-ecosystem/net-rpc-msgpackrpc2"
	"github.com/smallnest/rpcx/core"
	"github.com/smallnest/rpcx/log"
)

const (
	//DefaultRPCPath is the defaut HTTP RPC PATH
	DefaultRPCPath = "/_goRPC_"
)

var IP4Reg = regexp.MustCompile(`^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$`)

// ArgsContext contains net.Conn so services can get net.Conn info, for example, remote address.
type ArgsContext interface {
	Value(key string) interface{}
	SetValue(key string, value interface{})
}

type serverCodecWrapper struct {
	core.ServerCodec
	PluginContainer IServerPluginContainer
	Conn            net.Conn
	Timeout         time.Duration
	ReadTimeout     time.Duration
	WriteTimeout    time.Duration
}

// newServerCodecWrapper wraps a core.ServerCodec.
func newServerCodecWrapper(pc IServerPluginContainer, c core.ServerCodec, Conn net.Conn) *serverCodecWrapper {
	return &serverCodecWrapper{ServerCodec: c, PluginContainer: pc, Conn: Conn}
}

func (w *serverCodecWrapper) ReadRequestHeader(ctx context.Context, r *core.Request) error {
	if w.Timeout > 0 {
		w.Conn.SetDeadline(time.Now().Add(w.Timeout))
	}
	if w.ReadTimeout > 0 {
		w.Conn.SetReadDeadline(time.Now().Add(w.ReadTimeout))
	}

	// 设置client conn in context
	if m, ok := core.FromMapContext(ctx); ok {
		m[core.ConnKey] = w.Conn
	}

	//pre
	err := w.PluginContainer.DoPreReadRequestHeader(ctx, r)
	if err != nil {
		return err
	}
	err = w.ServerCodec.ReadRequestHeader(ctx, r)

	if err != nil {
		return err
	}

	//post
	err = w.PluginContainer.DoPostReadRequestHeader(ctx, r)
	return err
}

func (w *serverCodecWrapper) ReadRequestBody(ctx context.Context, body interface{}) error {
	//pre
	err := w.PluginContainer.DoPreReadRequestBody(ctx, body)
	if err != nil {
		return err
	}
	err = w.ServerCodec.ReadRequestBody(ctx, body)
	if err != nil {
		return err
	}

	if args, ok := body.(ArgsContext); ok {
		args.SetValue("conn", w.Conn)
	}

	//post
	err = w.PluginContainer.DoPostReadRequestBody(ctx, body)
	return err
}

func (w *serverCodecWrapper) WriteResponse(ctx context.Context, resp *core.Response, body interface{}) (err error) {
	if w.Timeout > 0 {
		w.Conn.SetDeadline(time.Now().Add(w.Timeout))
	}
	if w.ReadTimeout > 0 {
		w.Conn.SetWriteDeadline(time.Now().Add(w.WriteTimeout))
	}

	// pre
	if err = w.PluginContainer.DoPreWriteResponse(ctx, resp, body); err != nil {
		return
	}

	if err = w.ServerCodec.WriteResponse(ctx, resp, body); err != nil {
		return
	}

	// post
	return w.PluginContainer.DoPostWriteResponse(ctx, resp, body)
}

func (w *serverCodecWrapper) Close() (err error) {
	//pre
	err = w.ServerCodec.Close()
	//post

	return
}

// ServerCodecFunc is used to create a core.ServerCodec from net.Conn.
type ServerCodecFunc func(conn io.ReadWriteCloser) core.ServerCodec

// Server represents a RPC Server.
type Server struct {
	ServerCodecFunc ServerCodecFunc
	//PluginContainer must be configured before starting and Register plugins must be configured before invoking RegisterName method
	PluginContainer IServerPluginContainer
	//Metadata describes extra info about this service, for example, weight, active status
	Metadata     string
	rpcServer    *core.Server
	listener     net.Listener
	Timeout      time.Duration
	ReadTimeout  time.Duration
	WriteTimeout time.Duration
}

// NewServer returns a new Server.
func NewServer() *Server {
	return &Server{
		rpcServer:       core.NewServer(),
		PluginContainer: &ServerPluginContainer{plugins: make([]IPlugin, 0)},
		ServerCodecFunc: msgpackrpc2.NewServerCodec,
	}
}

// DefaultServer is the default instance of *Server.
var defaultServer = NewServer()

// Serve starts and listens RPC requests.
//It is blocked until receiving connectings from clients.
func Serve(n, address string) (err error) {
	return defaultServer.Serve(n, address)
}

// ServeTLS starts and listens RPC requests.
//It is blocked until receiving connectings from clients.
func ServeTLS(n, address string, config *tls.Config) (err error) {
	return defaultServer.ServeTLS(n, address, config)
}

// Start starts and listens RPC requests without blocking.
func Start(n, address string) (err error) {
	return defaultServer.Start(n, address)
}

// StartTLS starts and listens RPC requests without blocking.
func StartTLS(n, address string, config *tls.Config) (err error) {
	return defaultServer.StartTLS(n, address, config)
}

// ServeListener serve with a listener
func ServeListener(ln net.Listener) {
	defaultServer.ServeListener(ln)
}

// ServeByHTTP implements RPC via HTTP
func ServeByHTTP(ln net.Listener) {
	defaultServer.ServeByHTTP(ln, core.DefaultRPCPath)
}

// ServeByMux implements RPC via HTTP with customized mux
func ServeByMux(ln net.Listener, mux *http.ServeMux) {
	defaultServer.ServeByMux(ln, core.DefaultRPCPath, mux)
}

// SetServerCodecFunc sets a ServerCodecFunc
func SetServerCodecFunc(fn ServerCodecFunc) {
	defaultServer.ServerCodecFunc = fn
}

// Close closes RPC server.
func Close() error {
	return defaultServer.Close()
}

//GetListenedAddress return the listening address.
func GetListenedAddress() string {
	return defaultServer.Address()
}

// GetPluginContainer get PluginContainer of default server.
func GetPluginContainer() IServerPluginContainer {
	return defaultServer.PluginContainer
}

// RegisterName publishes in the server the set of methods .
func RegisterName(name string, service interface{}, metadata ...string) {
	defaultServer.RegisterName(name, service, metadata...)
}

// Auth sets authorization handler
func Auth(fn AuthorizationFunc) error {
	p := &AuthorizationServerPlugin{AuthorizationFunc: fn}
	return defaultServer.PluginContainer.Add(p)
}

func validIP4(ipAddress string) bool {
	ipAddress = strings.Trim(ipAddress, " ")
	i := strings.LastIndex(ipAddress, ":")
	ipAddress = ipAddress[:i] //remove port

	return IP4Reg.MatchString(ipAddress)
}

// Serve starts and listens RPC requests.
//It is blocked until receiving connectings from clients.
func (s *Server) Serve(network, address string) (err error) {
	var ln net.Listener
	ln, err = makeListener(network, address)
	if err != nil {
		return
	}

	s.listener = ln
	for {
		c, err := ln.Accept()
		if err != nil {
			continue
		}
		var ok bool
		if c, ok = s.PluginContainer.DoPostConnAccept(c); !ok {
			continue
		}

		wrapper := newServerCodecWrapper(s.PluginContainer, s.ServerCodecFunc(c), c)
		wrapper.Timeout = s.Timeout
		wrapper.ReadTimeout = s.ReadTimeout
		wrapper.WriteTimeout = s.WriteTimeout
		go s.rpcServer.ServeCodec(wrapper)
	}
}

// ServeTLS starts and listens RPC requests.
//It is blocked until receiving connectings from clients.
func (s *Server) ServeTLS(network, address string, config *tls.Config) (err error) {
	ln, err := tls.Listen(network, address, config)
	if err != nil {
		return
	}

	s.listener = ln
	for {
		c, err := ln.Accept()
		if err != nil {
			continue
		}

		var ok bool
		if c, ok = s.PluginContainer.DoPostConnAccept(c); !ok {
			continue
		}

		wrapper := newServerCodecWrapper(s.PluginContainer, s.ServerCodecFunc(c), c)
		wrapper.Timeout = s.Timeout
		wrapper.ReadTimeout = s.ReadTimeout
		wrapper.WriteTimeout = s.WriteTimeout
		go s.rpcServer.ServeCodec(wrapper)
	}
}

// ServeListener starts
func (s *Server) ServeListener(ln net.Listener) {
	s.listener = ln
	for {
		c, err := ln.Accept()
		if err != nil {
			continue
		}

		var ok bool
		if c, ok = s.PluginContainer.DoPostConnAccept(c); !ok {
			continue
		}

		wrapper := newServerCodecWrapper(s.PluginContainer, s.ServerCodecFunc(c), c)
		wrapper.Timeout = s.Timeout
		wrapper.ReadTimeout = s.ReadTimeout
		wrapper.WriteTimeout = s.WriteTimeout

		go s.rpcServer.ServeCodec(wrapper)
	}
}

// ServeByHTTP serves
func (s *Server) ServeByHTTP(ln net.Listener, rpcPath string) {
	http.Handle(rpcPath, s)
	srv := &http.Server{Handler: nil}
	srv.Serve(ln)
}

// ServeByMux serves
func (s *Server) ServeByMux(ln net.Listener, rpcPath string, mux *http.ServeMux) {
	mux.Handle(rpcPath, s)
	srv := &http.Server{Handler: mux}
	srv.Serve(ln)
}

var connected = "200 Connected to Go RPC"

//ServeHTTP implements net handler interface
func (s *Server) ServeHTTP(w http.ResponseWriter, req *http.Request) {
	if req.Method != "CONNECT" {
		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
		w.WriteHeader(http.StatusMethodNotAllowed)
		io.WriteString(w, "405 must CONNECT\n")
		return
	}
	c, _, err := w.(http.Hijacker).Hijack()
	if err != nil {
		log.Errorf("rpc hijacking %s : %v", req.RemoteAddr, err.Error())
		return
	}
	io.WriteString(c, "HTTP/1.0 "+connected+"\n\n")

	var ok bool
	if c, ok = s.PluginContainer.DoPostConnAccept(c); !ok {
		log.Errorf("client is not accepted: %s", c.RemoteAddr().String())
		return
	}

	wrapper := newServerCodecWrapper(s.PluginContainer, s.ServerCodecFunc(c), c)
	wrapper.Timeout = s.Timeout
	wrapper.ReadTimeout = s.ReadTimeout
	wrapper.WriteTimeout = s.WriteTimeout

	s.rpcServer.ServeCodec(wrapper)
}

// Start starts and listens RPC requests without blocking.
func (s *Server) Start(network, address string) (err error) {
	var ln net.Listener
	ln, err = makeListener(network, address)

	if err != nil {
		log.Errorf("failed to start server: %v", err)
		return
	}

	s.listener = ln

	go func() {
		for {
			c, err := ln.Accept()
			if err != nil {
				continue
			}

			var ok bool
			if c, ok = s.PluginContainer.DoPostConnAccept(c); !ok {
				log.Errorf("client is not accepted: %s", c.RemoteAddr().String())
				continue
			}
			wrapper := newServerCodecWrapper(s.PluginContainer, s.ServerCodecFunc(c), c)
			wrapper.Timeout = s.Timeout
			wrapper.ReadTimeout = s.ReadTimeout
			wrapper.WriteTimeout = s.WriteTimeout

			go s.rpcServer.ServeCodec(wrapper)
		}
	}()

	return
}

// StartTLS starts and listens RPC requests without blocking.
func (s *Server) StartTLS(network, address string, config *tls.Config) (err error) {
	ln, err := tls.Listen(network, address, config)
	if err != nil {
		log.Errorf("failed to start server: %v", err)
		return
	}

	s.listener = ln

	go func() {
		for {
			c, err := ln.Accept()
			if err != nil {
				continue
			}

			var ok bool
			if c, ok = s.PluginContainer.DoPostConnAccept(c); !ok {
				log.Errorf("client is not accepted: %s", c.RemoteAddr().String())
				continue
			}
			wrapper := newServerCodecWrapper(s.PluginContainer, s.ServerCodecFunc(c), c)
			wrapper.Timeout = s.Timeout
			wrapper.ReadTimeout = s.ReadTimeout
			wrapper.WriteTimeout = s.WriteTimeout

			go s.rpcServer.ServeCodec(wrapper)
		}
	}()

	return
}

// Close closes RPC server.
func (s *Server) Close() error {
	return s.listener.Close()
}

// Address return the listening address.
func (s *Server) Address() string {
	return s.listener.Addr().String()
}

// RegisterName publishes in the server the set of methods of the
// receiver value that satisfy the following conditions:
//	- exported method of exported type
//	- two arguments, both of exported type
//	- the second argument is a pointer
//	- one return value, of type error
// It returns an error if the receiver is not an exported type or has
// no suitable methods. It also logs the error using package log.
// The client accesses each method using a string of the form "Type.Method",
// where Type is the receiver's concrete type.
func (s *Server) RegisterName(name string, service interface{}, metadata ...string) {
	s.rpcServer.RegisterName(name, service)
	s.PluginContainer.DoRegister(name, service, metadata...)
}

//Auth sets authorization function
func (s *Server) Auth(fn AuthorizationFunc) error {
	p := &AuthorizationServerPlugin{AuthorizationFunc: fn}
	return s.PluginContainer.Add(p)
}
